rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,315 @@
+- RFC PR: <!-- leave this empty, to be filled in later -->
+- Related Issues: <!-- if relevant -->
+- Authors: Nicolò Ribaudo
+- Champion: Nicolò Ribaudo
+- Implementors: <!-- the names of everyone who will work on the PR. you can leave this empty if you would like someone else to work on it -->
+
+# Summary
+
+Many of our plugins have a `loose` option, which enables the compiler to make some assumptions about the code you are writing. It uses these to ignore certain edge cases and generate a smaller or faster output.
+
+The `loose` option has been around since `6to5`: at first as a top level option, and then it was split into each plugin as a catch all for anything that we wanted to be less spec compliant.
+
+However, `loose` has some problems: namely that the term itself is not clear what it means, and _should_ be more related to a part of the language rather than to the plugin that it's compiling it (see the "Motivation" section for more details). A lot of this behavior is documented with a single example, so in the future we’ll want to be specific with the assumptions themselves and maybe provide a set of tests/codesandboxes explaining differences.
+
+This RFC proposes introducing a new *top-level* option, `"assumptions"`, which is an object containing different flags which can simplify the output code.
+These flags have two characteristics:
+- They specify _something_ about a specific language feature, and not about a specific Babel plugin.
+- Different flags can toggle different unrelated optimizations in the same plugin, instead of toggling both of them with a single generic `"loose"` option.
+
+# Basic example
+
+```jsonc
+// babel.config.json
+{
+  "assumptions": {
+    "noDocumentAll": true,
+    "noGetterSideEffects": true,
+    "iterableIsArray": true
+  },
+  "presets": ["@babel/preset-env"]
+}
+```
+
+<details>
+<summary>Input / Output</summary>
+
+```javascript 
+// input code
+
+for (const el of arr) {
+  el.logger?.log("Ok!");
+}
+```
+
+```javascript 
+// output code
+
+for (let i = 0; i < arr.length; i++) {
+  const el = arr[i];
+  el.logger == null ? void 0 : el.logger.log("Ok!");
+}
+```
+</details>
+<br />
+
+```jsonc
+// babel.config.json
+{
+  "assumptions": {
+    "noDocumentAll": false,
+    "noGetterSideEffects": true,
+    "iterableIsArray": true
+  },
+  "presets": ["@babel/preset-env"]
+}
+```
+
+
+<details>
+<summary>Input / Output</summary>
+
+```javascript 
+// input code
+
+for (const el of arr) {
+  el.logger?.log("Ok!");
+}
+```
+
+```javascript 
+// output code
+
+for (let i = 0; i < arr.length; i++) {
+  const el = arr[i];
+  el.logger === null || el.logger === void 0 ? void 0 : el.logger.log("Ok!");
+}
+```
+</details>
+
+# Motivation
+
+The `loose` options have different problems:
+- _"What is loose?"_ It's not always clear what assumptions `loose` makes. You always have to check in the docs, end even the docs often only partially mention what `loose` means in that specific context.
```

Comment:
```suggestion
- _"What is loose?"_ Since `loose` is not descriptive, it requires users to check a plugin's docs to see what assumptions it makes (and the docs often only partially mention what `loose` means in a specific context).
```